src/common/subprocess/subprocess.h

```diff
@@ -0,0 +1,370 @@
+ *   called from a callback set in 'on_channel_out'.
+ *
+ *  The `name` argument is also used as the name of an environment variable
+ *   in the subprocess environment that is set to the file descriptor number
+ *   of the process side of the socketpair. E.g. name = "FLUX_PMI" would
+ *   result in the environment variable "FLUX_PMI_FD=N" set in the process
+ *   environment.
+ */
+int flux_cmd_add_channel (flux_cmd_t *cmd, const char *name);
+
+/*
+ *  Set generic string options for command object `cmd`. As with environment
+ *   variables, this function adds the option `var` to with value `val` to
+ *   the options array for this command. This can be used to enable optional
+ *   behavior for executed processes (e.g. setpgrp(2))
+ *
+ *  String options, note that name indicates the 'name' argument used
+ *  in flux_cmd_add_channel() above.
+ *
+ *  name + "_BUFSIZE" = buffer size
+ *  STDIN_BUFSIZE = buffer size
+ *  STDOUT_BUFSIZE = buffer size
+ *  STDERR_BUFSIZE = buffer size
+ *
+ *  By default, stdio and channels use an internal buffer of 1 meg.
+ *  The buffer size can be adjusted with this option.
+ */
+int flux_cmd_setopt (flux_cmd_t *cmd, const char *var, const char *val);
+const char *flux_cmd_getopt (flux_cmd_t *cmd, const char *var);
+
+
+
+/*
+ *  Subprocesses:
+ */
+
+/*
+ *  Asynchronously create a new subprocess described by command object
+ *   `cmd`.  flux_exec() and flux_local_exec() create a new subprocess
+ *   locally.  flux_rexec() creates a new subprocess on Flux rank
+ *   `rank`. Callbacks in `ops` structure that are non-NULL will be
+ *   called to process state changes, I/O, and completion.
+ *
+ *  'rank' can be set to FLUX_NODEID_ANY or FLUX_NODEID_UPSTREAM.
+ *
+ *  This function may return NULL (with errno set) on invalid
+ *   argument(s) (EINVAL), or failure of underlying Flux messaging
+ *   calls. Otherwise, a valid subprocess object is returned, though
+ *   there is no guarantee the subprocess has started running anywhere
+ *   by the time the call returns.
+ *
+ */
+flux_subprocess_t *flux_exec (flux_t *h, int flags,
+                              const flux_cmd_t *cmd,
+                              flux_subprocess_ops_t *ops);
+
+flux_subprocess_t *flux_local_exec (flux_reactor_t *r, int flags,
+                                    const flux_cmd_t *cmd,
+                                    flux_subprocess_ops_t *ops);
+
+flux_subprocess_t *flux_rexec (flux_t *h, int rank, int flags,
+                               const flux_cmd_t *cmd,
+                               flux_subprocess_ops_t *ops);
+
+
+/*
+ *  Write data to "stream" stream of subprocess `p`.  'stream' can be
+ *  "STDIN" or the name of a stream specified with
+ *  flux_cmd_add_channel().  If 'stream' is NULL, defaults to "STDIN".
+ *
+ *  Returns the total amount of data successfully buffered.
+ */
+int flux_subprocess_write (flux_subprocess_t *p, const char *stream,
+                           const char *buf, size_t len);
+
```

Comment:
Do stream/channel names have to be all CAPS? It would be nice if "stdin" "stdout" "stderr" worked as well to match the `FILE *` stream names.

Reply:
No, they don't have to be in all CAPS.  I wanted to differentiate the stream name from the ```FILE *``` stream (i.e. when searching in the code).

Reply:
ok, great point! :-)